lib/slashdeploy/auth.rb

```diff
@@ -31,11 +31,17 @@ def call(env)
 
       begin
         handler.call(env)
-      rescue User::MissingGitHubAccount
-        # If we don't know this slack user, we'll ask them to authenticate
-        # with GitHub. We encode and sign the Slack user id within the state
-        # param so we know what slack user they are when the hit the GitHub
-        # callback.
+      rescue User::MissingGitHubAccount, Octokit::Unauthorized
```

Comment:
Doing this so broadly may end up having some unintended side effects. For example, if the access token is valid, but you don't have access to a specific repository, this is the exception that's raised.

We should probably have a way to "test" the token first, to verify that it's actually invalid, probably using the [`GET /user` api](https://developer.github.com/v3/users/#get-the-authenticated-user). That will let us know 100% that the token is invalid if we get a 401:

```console
$ curl https://api.github.com/user -i
HTTP/1.1 401 Unauthorized
Date: Thu, 19 Apr 2018 01:24:35 GMT
Content-Type: application/json; charset=utf-8
Content-Length: 135
Server: GitHub.com
Status: 401 Unauthorized

{
  "message": "Requires authentication",
  "documentation_url": "https://developer.github.com/v3/users/#get-the-authenticated-user"
}
```

Reply:
Octokit raises a different exception when access to the repo is unauthorized. Catching this exception is catching anything related to `Authentication` (passwords, tokens, etc) and it is an `HTTP 401`.

Any errors related to `Authorization` (like being authenticated but not allowed to view a particular repo) will throw an `HTTP 403`.

Reference: https://github.com/octokit/octokit.rb/blob/master/lib/octokit/error.rb#L238

Reply:
I read through every spec and this seems like the best place for it, but I'm very new to this repo.

I was thinking maybe you could put the test here.
https://github.com/remind101/slashdeploy/blob/master/spec/controllers/sessions_controller_spec.rb